kyaml/yaml/types.go

```diff
@@ -44,16 +44,17 @@ func IsMissingOrNull(node *RNode) bool {
 	return node == nil || node.YNode() == nil || node.YNode().Tag == NullNodeTag
 }
 
-// IsEmpty returns true if the RNode is MissingOrNull, or is either a MappingNode with
-// no fields.
+// IsEmpty returns true if the RNode is MissingOrNull
 func IsEmpty(node *RNode) bool {
```

Comment:
It now seems we have:

```
IsMissingOrNull(n) => n == nil || n.YNode() == nil || n.YNode().Tag == NullNodeTag
        IsEmpty(n) => IsMissingOrNull(n)
         IsNull(n) => n != nil && n.YNode() != nil && n.YNode().Tag == NullNodeTag
     IsEmptyMap(n) => IsEmpty(n) ||
                          (n.YNode().Kind == yaml.MappingNode && len(n.YNode().Content) == 0)
```

This feels random.  And we have this strange contest between
`Kind` and `Tag`, which i'm not sure how to resolve yet.

We've not hit 1.0 yet, so we can change the API.

How about we try:

```
  IsNil(n)        => n == nil || n.YNode() == nil
  IsEmptyDoc(n)   => !IsNodeNil(n) && n.YNode().Tag == NodeTagNull
  IsEmptyMap(n)   => !IsNodeNil(n) &&
                        n.YNode().Kind == yaml.MappingNode &&
                        len(n.YNode().Content) == 0
```

with this, the tags are hidden, and we start a pattern where
`IsEmptyFoo` means the object is non-nil with a defined type -
so empty docs, lists and maps are different things.

Translation:

 - `IsNull` becomes `IsEmptyDoc`
 - `IsEmpty`and `IsMissingOrNull` are written as `IsNil || IsEmptyDoc`


The compiler can optimize away redundancy.


So L54 in `kyaml/yaml/walk/associative_sequence.go` which currently reads

```
if yaml.IsEmpty(val) || yaml.IsEmptyMap(val) {
```

becomes

```
if yaml.IsNil(val) || yaml.IsEmptyDoc(val) || yaml.IsEmptyMap(val) {

```

Reply:
That sounds good. But because these methods are fundamental and widely used so if we change them, we need to update a lot of files.

Reply:
What about create another issue about this refactoring?